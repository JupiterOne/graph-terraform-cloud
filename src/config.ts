import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationProviderAuthenticationError,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { TerraformCloudClient } from './tfe/client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  apiKey: {
    type: 'string',
    mask: true,
  },
  organizationName: {
    type: 'string',
  },
};

export interface IntegrationConfig extends IntegrationInstanceConfig {
  apiKey: string;
  organizationName: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { instance } = context;
  const { config } = instance;

  if (!config.apiKey || !config.organizationName) {
    throw new IntegrationValidationError(
      'Config requires all of {apiKey, organizationName}',
    );
  }

  const client = new TerraformCloudClient({ apiKey: config.apiKey });

  try {
    await client.organizations.list();
  } catch (err) {
    throw new IntegrationProviderAuthenticationError(err.message);
  }
}
