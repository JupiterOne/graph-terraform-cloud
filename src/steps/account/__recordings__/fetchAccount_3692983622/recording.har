{
  "log": {
    "_recordingName": "fetchAccount",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "020c85ae52b49216a932f1ac0762939b",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/vnd.api+json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.terraform.io"
            }
          ],
          "headersSize": 373,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.terraform.io/api/v2/account/details"
        },
        "response": {
          "bodySize": 823,
          "content": {
            "mimeType": "application/vnd.api+json; charset=utf-8",
            "size": 823,
            "text": "{\"data\":{\"id\":\"user-TSVGDCJGAaCZs8T7\",\"type\":\"users\",\"attributes\":{\"username\":\"api-org-jupiterone-HXJbgND3Bo\",\"is-service-account\":true,\"avatar-url\":\"https://www.gravatar.com/avatar/c726702df6456d9affb1eb8d9d37a9de?s=100\\u0026d=mm\",\"password\":null,\"enterprise-support\":false,\"is-site-admin\":false,\"is-sso-login\":false,\"two-factor\":{\"enabled\":false,\"verified\":false},\"email\":\"api-org-jupiterone-hxjbgnd3bo@hashicorp.com\",\"unconfirmed-email\":null,\"has-git-hub-app-token\":false,\"is-confirmed\":true,\"is-sudo\":false,\"permissions\":{\"can-create-organizations\":false,\"can-change-email\":true,\"can-change-username\":true,\"can-manage-user-tokens\":false}},\"relationships\":{\"authentication-tokens\":{\"links\":{\"related\":\"/api/v2/users/user-TSVGDCJGAaCZs8T7/authentication-tokens\"}}},\"links\":{\"self\":\"/api/v2/users/user-TSVGDCJGAaCZs8T7\"}}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 02 Feb 2022 12:38:52 GMT"
            },
            {
              "name": "content-type",
              "value": "application/vnd.api+json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "823"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "etag",
              "value": "W/\"d26a1432ec06c54c79f856ace5deeb0a\""
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "tfp-api-version",
              "value": "2.5"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Origin"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-download-options",
              "value": "noopen"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-permitted-cross-domain-policies",
              "value": "none"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "30"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "29"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "0.199"
            },
            {
              "name": "x-request-id",
              "value": "b0cbf703-ec99-3f82-e55c-d6bdf15c9627"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            }
          ],
          "headersSize": 698,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-02T12:38:52.266Z",
        "time": 597,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 597
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
