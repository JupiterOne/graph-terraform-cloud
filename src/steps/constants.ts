import {
  RelationshipClass,
  StepEntityMetadata,
  StepRelationshipMetadata,
} from '@jupiterone/integration-sdk-core';

export const ACCOUNT_ENTITY_KEY = 'ACCOUNT_ENTITY';
export const ACCOUNT_DATA_KEY = 'ACCOUNT_DATA';
export const MAC_OS_CONFIGURATION_DETAILS_BY_ID_KEY =
  'MAC_OS_CONFIGURATION_DETAILS_BY_ID';

export enum IntegrationSteps {
  ACCOUNT = 'fetch-account',
  ORGANIZATIONS = 'fetch-organizations',
  ORGANIZATION_ENTITLEMENT_SET = 'fetch-organization-entitlement-set',
  ORGANIZATION_MEMBERS = 'fetch-organization-members',
  ORGANIZATION_OAUTH_TOKENS = 'fetch-organization-oauth-tokens',
  ORGANIZATION_WORKSPACES = 'fetch-organization-workspaces',
  ORGANIZATION_TEAMS = 'fetch-organization-teams',
  WORKSPACE_RESOURCES = 'fetch-workspace-resources',
}

export const Entities: Record<
  | 'ACCOUNT'
  | 'ENTITLEMENT_SET'
  | 'ORGANIZATION'
  | 'TEAM'
  | 'USER'
  | 'WORKSPACE'
  | 'WORKSPACE_RESOURCE',
  StepEntityMetadata
> = {
  ACCOUNT: {
    _type: 'tfe_account',
    _class: ['Account'],
    resourceName: 'Account',
  },
  ENTITLEMENT_SET: {
    _type: 'tfe_entitlement_set',
    _class: ['Entity'],
    resourceName: 'Entitlement Set',
  },
  ORGANIZATION: {
    _type: 'tfe_organization',
    _class: ['Organization'],
    resourceName: 'Organization',
  },
  TEAM: {
    _type: 'tfe_team',
    _class: ['Team'],
    resourceName: 'Team',
  },
  USER: {
    _type: 'tfe_user',
    _class: ['User'],
    resourceName: 'User',
  },
  WORKSPACE: {
    _type: 'tfe_workspace',
    _class: ['Project'],
    resourceName: 'Workspace',
  },
  WORKSPACE_RESOURCE: {
    _type: 'tfe_workspace_resource',
    _class: ['Resource'],
    resourceName: 'Resource',
  },
};

export const Relationships: Record<
  | 'ACCOUNT_HAS_ORGANIZATION'
  | 'ORGANIZATION_HAS_ENTITLEMENT_SET'
  | 'ORGANIZATION_HAS_TEAM'
  | 'ORGANIZATION_HAS_USER'
  | 'ORGANIZATION_HAS_WORKSPACE'
  | 'WORKSPACE_HAS_RESOURCE',
  StepRelationshipMetadata
> = {
  ACCOUNT_HAS_ORGANIZATION: {
    _type: 'tfe_account_has_organization',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ACCOUNT._type,
    targetType: Entities.ORGANIZATION._type,
  },
  ORGANIZATION_HAS_ENTITLEMENT_SET: {
    _type: 'tfe_organization_has_entitlement_set',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ORGANIZATION._type,
    targetType: Entities.ENTITLEMENT_SET._type,
  },
  ORGANIZATION_HAS_TEAM: {
    _type: 'tfe_organization_has_team',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ORGANIZATION._type,
    targetType: Entities.TEAM._type,
  },
  ORGANIZATION_HAS_USER: {
    _type: 'tfe_organization_has_user',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ORGANIZATION._type,
    targetType: Entities.USER._type,
  },
  ORGANIZATION_HAS_WORKSPACE: {
    _type: 'tfe_organization_has_workspace',
    _class: RelationshipClass.HAS,
    sourceType: Entities.ORGANIZATION._type,
    targetType: Entities.WORKSPACE._type,
  },
  WORKSPACE_HAS_RESOURCE: {
    _type: 'tfe_workspace_has_resource',
    _class: RelationshipClass.HAS,
    sourceType: Entities.WORKSPACE._type,
    targetType: Entities.WORKSPACE_RESOURCE._type,
  },
};
