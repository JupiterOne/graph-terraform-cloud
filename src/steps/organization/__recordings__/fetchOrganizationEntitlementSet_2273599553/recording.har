{
  "log": {
    "_recordingName": "fetchOrganizationEntitlementSet",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "f23a3cfcb0c06e6dd84611fcfcb09bd6",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/vnd.api+json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.terraform.io"
            }
          ],
          "headersSize": 398,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.terraform.io/api/v2/organizations/jupiterone/entitlement-set"
        },
        "response": {
          "bodySize": 439,
          "content": {
            "mimeType": "application/vnd.api+json; charset=utf-8",
            "size": 439,
            "text": "{\"data\":{\"id\":\"org-yMai1ZUhS39WMAU9\",\"type\":\"entitlement-sets\",\"attributes\":{\"cost-estimation\":false,\"configuration-designer\":true,\"operations\":true,\"private-module-registry\":true,\"sentinel\":false,\"state-storage\":true,\"teams\":false,\"vcs-integrations\":true,\"usage-reporting\":false,\"user-limit\":5.0,\"self-serve-billing\":true,\"audit-logging\":false,\"agents\":false,\"sso\":false},\"links\":{\"self\":\"/api/v2/entitlement-sets/org-yMai1ZUhS39WMAU9\"}}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 31 Dec 2021 03:21:11 GMT"
            },
            {
              "name": "content-type",
              "value": "application/vnd.api+json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "439"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "etag",
              "value": "W/\"c73baa90b1c2b209b04c0729babf3ffb\""
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "tfp-api-version",
              "value": "2.5"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Origin"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-download-options",
              "value": "noopen"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-permitted-cross-domain-policies",
              "value": "none"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "30"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "29"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "0.06"
            },
            {
              "name": "x-request-id",
              "value": "8af0bff3-c315-c660-fcda-0c9adc643304"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            }
          ],
          "headersSize": 697,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-12-31T03:21:11.044Z",
        "time": 958,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 958
        }
      },
      {
        "_id": "3d52198674d63c5f240bad80c07a6e94",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/vnd.api+json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.terraform.io"
            }
          ],
          "headersSize": 399,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [],
          "url": "https://app.terraform.io/api/v2/organizations/j1-test-org/entitlement-set"
        },
        "response": {
          "bodySize": 437,
          "content": {
            "mimeType": "application/vnd.api+json; charset=utf-8",
            "size": 437,
            "text": "{\"data\":{\"id\":\"org-qcDVXfQDYmRkaSDr\",\"type\":\"entitlement-sets\",\"attributes\":{\"cost-estimation\":true,\"configuration-designer\":true,\"operations\":true,\"private-module-registry\":true,\"sentinel\":true,\"state-storage\":true,\"teams\":true,\"vcs-integrations\":true,\"usage-reporting\":false,\"user-limit\":null,\"self-serve-billing\":true,\"audit-logging\":false,\"agents\":false,\"sso\":false},\"links\":{\"self\":\"/api/v2/entitlement-sets/org-qcDVXfQDYmRkaSDr\"}}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Fri, 31 Dec 2021 03:24:49 GMT"
            },
            {
              "name": "content-type",
              "value": "application/vnd.api+json; charset=utf-8"
            },
            {
              "name": "content-length",
              "value": "437"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "etag",
              "value": "W/\"1ab3ed7148849be49e2a1cca6372c37c\""
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "tfp-api-version",
              "value": "2.5"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Origin"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-download-options",
              "value": "noopen"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-permitted-cross-domain-policies",
              "value": "none"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "30"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "29"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "0.459"
            },
            {
              "name": "x-request-id",
              "value": "f7b46b23-a65a-5f4e-df55-784cc85d5d98"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            }
          ],
          "headersSize": 698,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-12-31T03:24:48.644Z",
        "time": 984,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 984
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
