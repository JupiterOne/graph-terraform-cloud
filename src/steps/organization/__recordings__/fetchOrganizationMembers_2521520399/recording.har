{
  "log": {
    "_recordingName": "fetchOrganizationMembers",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "f7815dab5838a3db65aed5c578ca0d2e",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/vnd.api+json"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "*/*"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "app.terraform.io"
            }
          ],
          "headersSize": 439,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "page",
              "value": {
                "size": "100"
              }
            },
            {
              "name": "include",
              "value": "user"
            }
          ],
          "url": "https://app.terraform.io/api/v2/organizations/jupiterone/organization-memberships?page%5Bsize%5D=100&include=user"
        },
        "response": {
          "bodySize": 4649,
          "content": {
            "mimeType": "application/vnd.api+json; charset=utf-8",
            "size": 4649,
            "text": "{\"data\":[{\"id\":\"ou-K9x7j5x3iJ4wMV4j\",\"type\":\"organization-memberships\",\"attributes\":{\"status\":\"active\",\"email\":\"austin.kelleher@jupiterone.com\",\"created-at\":\"2021-02-17T21:24:28.951Z\"},\"relationships\":{\"teams\":{\"data\":[{\"id\":\"team-iPT3DfbPpZpEjb9t\",\"type\":\"teams\"}]},\"user\":{\"data\":{\"id\":\"user-dtoAE5LXupfvycjK\",\"type\":\"users\"}},\"organization\":{\"data\":{\"id\":\"jupiterone\",\"type\":\"organizations\"}}}},{\"id\":\"ou-eKePhn8G3wmb3skq\",\"type\":\"organization-memberships\",\"attributes\":{\"status\":\"active\",\"email\":\"nick.dowmon@jupiterone.com\",\"created-at\":\"2021-03-17T18:16:04.684Z\"},\"relationships\":{\"teams\":{\"data\":[{\"id\":\"team-iPT3DfbPpZpEjb9t\",\"type\":\"teams\"}]},\"user\":{\"data\":{\"id\":\"user-HMyUEwhb3mB45HZ7\",\"type\":\"users\"}},\"organization\":{\"data\":{\"id\":\"jupiterone\",\"type\":\"organizations\"}}}},{\"id\":\"ou-aEtsGQRPTSHtnx43\",\"type\":\"organization-memberships\",\"attributes\":{\"status\":\"active\",\"email\":\"pgilaga@gmail.com\",\"created-at\":\"2021-07-15T20:50:55.220Z\"},\"relationships\":{\"teams\":{\"data\":[{\"id\":\"team-iPT3DfbPpZpEjb9t\",\"type\":\"teams\"}]},\"user\":{\"data\":{\"id\":\"user-2yZJWZGNJmNsidGK\",\"type\":\"users\"}},\"organization\":{\"data\":{\"id\":\"jupiterone\",\"type\":\"organizations\"}}}},{\"id\":\"ou-H85Ft3VRpgP8JHux\",\"type\":\"organization-memberships\",\"attributes\":{\"status\":\"active\",\"email\":\"stefan@creativice.com\",\"created-at\":\"2021-02-18T00:47:35.848Z\"},\"relationships\":{\"teams\":{\"data\":[{\"id\":\"team-iPT3DfbPpZpEjb9t\",\"type\":\"teams\"}]},\"user\":{\"data\":{\"id\":\"user-qZQNfnC19eG6Ameo\",\"type\":\"users\"}},\"organization\":{\"data\":{\"id\":\"jupiterone\",\"type\":\"organizations\"}}}}],\"included\":[{\"id\":\"user-dtoAE5LXupfvycjK\",\"type\":\"users\",\"attributes\":{\"username\":\"austinkelleher-jupiterone\",\"is-service-account\":false,\"avatar-url\":\"https://www.gravatar.com/avatar/1941eaaff412c1a04cfd4166bfd66371?s=100\\u0026d=mm\",\"two-factor\":{\"enabled\":false,\"verified\":false},\"email\":\"austin.kelleher@jupiterone.com\",\"permissions\":{\"can-create-organizations\":false,\"can-change-email\":true,\"can-change-username\":true,\"can-manage-user-tokens\":false}},\"relationships\":{\"authentication-tokens\":{\"links\":{\"related\":\"/api/v2/users/user-dtoAE5LXupfvycjK/authentication-tokens\"}}},\"links\":{\"self\":\"/api/v2/users/user-dtoAE5LXupfvycjK\"}},{\"id\":\"user-HMyUEwhb3mB45HZ7\",\"type\":\"users\",\"attributes\":{\"username\":\"ndowmon\",\"is-service-account\":false,\"avatar-url\":\"https://www.gravatar.com/avatar/4cdbb2e21e2d991b0dd11d9a38c2c660?s=100\\u0026d=mm\",\"two-factor\":{\"enabled\":false,\"verified\":false},\"email\":\"nick.dowmon@jupiterone.com\",\"permissions\":{\"can-create-organizations\":false,\"can-change-email\":true,\"can-change-username\":true,\"can-manage-user-tokens\":false}},\"relationships\":{\"authentication-tokens\":{\"links\":{\"related\":\"/api/v2/users/user-HMyUEwhb3mB45HZ7/authentication-tokens\"}}},\"links\":{\"self\":\"/api/v2/users/user-HMyUEwhb3mB45HZ7\"}},{\"id\":\"user-2yZJWZGNJmNsidGK\",\"type\":\"users\",\"attributes\":{\"username\":\"pgilaga\",\"is-service-account\":false,\"avatar-url\":\"https://www.gravatar.com/avatar/6bf8d79d3b1d6d34a9d69f62cd09d20a?s=100\\u0026d=mm\",\"two-factor\":{\"enabled\":false,\"verified\":false},\"email\":\"pgilaga@gmail.com\",\"permissions\":{\"can-create-organizations\":false,\"can-change-email\":true,\"can-change-username\":true,\"can-manage-user-tokens\":false}},\"relationships\":{\"authentication-tokens\":{\"links\":{\"related\":\"/api/v2/users/user-2yZJWZGNJmNsidGK/authentication-tokens\"}}},\"links\":{\"self\":\"/api/v2/users/user-2yZJWZGNJmNsidGK\"}},{\"id\":\"user-qZQNfnC19eG6Ameo\",\"type\":\"users\",\"attributes\":{\"username\":\"stefan-creativice\",\"is-service-account\":false,\"avatar-url\":\"https://www.gravatar.com/avatar/5a764b0f770a7bfe74cdc94afea04ed6?s=100\\u0026d=mm\",\"two-factor\":{\"enabled\":false,\"verified\":false},\"email\":\"stefan@creativice.com\",\"permissions\":{\"can-create-organizations\":false,\"can-change-email\":true,\"can-change-username\":true,\"can-manage-user-tokens\":false}},\"relationships\":{\"authentication-tokens\":{\"links\":{\"related\":\"/api/v2/users/user-qZQNfnC19eG6Ameo/authentication-tokens\"}}},\"links\":{\"self\":\"/api/v2/users/user-qZQNfnC19eG6Ameo\"}}],\"links\":{\"self\":\"https://app.terraform.io/api/v2/organizations/jupiterone/organization-memberships?include=user\\u0026page%5Bnumber%5D=1\\u0026page%5Bsize%5D=100\",\"first\":\"https://app.terraform.io/api/v2/organizations/jupiterone/organization-memberships?include=user\\u0026page%5Bnumber%5D=1\\u0026page%5Bsize%5D=100\",\"prev\":null,\"next\":null,\"last\":\"https://app.terraform.io/api/v2/organizations/jupiterone/organization-memberships?include=user\\u0026page%5Bnumber%5D=1\\u0026page%5Bsize%5D=100\"},\"meta\":{\"status-counts\":{\"total\":4,\"matching\":4,\"active\":4,\"invited\":0},\"pagination\":{\"current-page\":1,\"page-size\":100,\"prev-page\":null,\"next-page\":null,\"total-pages\":1,\"total-count\":4}}}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "date",
              "value": "Wed, 02 Feb 2022 12:40:44 GMT"
            },
            {
              "name": "content-type",
              "value": "application/vnd.api+json; charset=utf-8"
            },
            {
              "name": "transfer-encoding",
              "value": "chunked"
            },
            {
              "name": "connection",
              "value": "close"
            },
            {
              "name": "cache-control",
              "value": "max-age=0, private, must-revalidate"
            },
            {
              "name": "etag",
              "value": "W/\"0a231a2d1b62689ad085bf12a777a5e2\""
            },
            {
              "name": "referrer-policy",
              "value": "strict-origin-when-cross-origin"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000; includeSubDomains; preload"
            },
            {
              "name": "tfp-api-version",
              "value": "2.5"
            },
            {
              "name": "vary",
              "value": "Accept-Encoding, Origin"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-download-options",
              "value": "noopen"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-permitted-cross-domain-policies",
              "value": "none"
            },
            {
              "name": "x-ratelimit-limit",
              "value": "30"
            },
            {
              "name": "x-ratelimit-remaining",
              "value": "29"
            },
            {
              "name": "x-ratelimit-reset",
              "value": "0.51"
            },
            {
              "name": "x-request-id",
              "value": "a3e45e1c-33fd-a83c-42a5-a7461c27b76f"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            }
          ],
          "headersSize": 704,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2022-02-02T12:40:43.935Z",
        "time": 651,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 651
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
