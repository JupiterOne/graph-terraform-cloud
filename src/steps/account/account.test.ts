import { fetchAccount } from '.';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities } from '../constants';
import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import { IntegrationConfig } from '../../config';

describe('#fetchAccount', () => {
  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await createDataCollectionTest({
      context,
      recordingName: 'fetchAccount',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchAccount],
      entitySchemaMatchers: [
        {
          _type: Entities.ACCOUNT._type,
          matcher: {
            _class: ['Account'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'tfe_account' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                email: { type: 'string' },
                username: { type: 'string' },
              },
            },
          },
        },
      ],
    });
  });
});
