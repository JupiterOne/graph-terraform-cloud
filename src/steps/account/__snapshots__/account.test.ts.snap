// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchAccount should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "account-user-TSVGDCJGAaCZs8T7",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "avatarUrl": "https://www.gravatar.com/avatar/c726702df6456d9affb1eb8d9d37a9de?s=100&d=mm",
            "email": "api-org-jupiterone-hxjbgnd3bo@hashicorp.com",
            "enterpriseSupport": false,
            "hasGitHubAppToken": false,
            "id": "user-TSVGDCJGAaCZs8T7",
            "isConfirmed": true,
            "isServiceAccount": true,
            "isSiteAdmin": false,
            "isSsoLogin": false,
            "isSudo": false,
            "password": null,
            "permissions": Object {
              "canChangeEmail": true,
              "canChangeUsername": true,
              "canCreateOrganizations": false,
              "canManageUserTokens": false,
            },
            "twoFactor": Object {
              "enabled": false,
              "verified": false,
            },
            "unconfirmedEmail": null,
            "username": "api-org-jupiterone-HXJbgND3Bo",
          },
        },
      ],
      "_type": "tfe_account",
      "createdOn": undefined,
      "displayName": "api-org-jupiterone-HXJbgND3Bo",
      "email": "api-org-jupiterone-hxjbgnd3bo@hashicorp.com",
      "id": "user-TSVGDCJGAaCZs8T7",
      "name": "api-org-jupiterone-HXJbgND3Bo",
      "username": "api-org-jupiterone-HXJbgND3Bo",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "tfe_account",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
